pymongo/uri_parser.py

```diff
@@ -34,26 +40,49 @@
 SRV_SCHEME_LEN = len(SRV_SCHEME)
 DEFAULT_PORT = 27017
 
+# List of sub-delimiters as defined in RFC 3986.
+SUBDELIMS = ["!", "$", "&", "'", "(", ")", "*", "+", ",", ";", "="]
+
 
 def parse_userinfo(userinfo):
     """Validates the format of user information in a MongoDB URI.
-    Reserved characters like ':', '/', '+' and '@' must be escaped
-    following RFC 3986.
+    Reserved characters that are gen-delimiters (":", "/", "?", "#", "[",
+    "]", "@") as per RFC 3986 must be escaped.
 
     Returns a 2-tuple containing the unescaped username followed
     by the unescaped password.
 
     :Paramaters:
         - `userinfo`: A string of the form <username>:<password>
+
+    .. versionchanged:: 4.0
+       Reserved characters that are sub-delimiters ("!", "$", "&", "'",
+       "(", ")", "*", "+", ",", ";", "=") as per RFC 3986 need not be
+       escaped.
     """
     if '@' in userinfo or userinfo.count(':') > 1:
```

Comment:
I think these checks are redundant now. The new check will cover these cases (`@` and `:`) too.

Reply:
I don't think these checks are redundant because we don't check for the number of colons. For example, with those checks removed `parse_userinfo("user:pass:pass2")` does not raise an error, which it should.